From b281a10c96f065731f869dd3b9abe9d6cb8973c5 Mon Sep 17 00:00:00 2001
From: David Bauer <mail@david-bauer.net>
Date: Mon, 26 Jan 2026 23:03:07 +0100
Subject: [PATCH] wifi: mt76: don't return TXQ when exceeding max non-AQL
 packets

mt76_txq_send_burst does check if the number of non-AQL frames exceeds
the maximum. In this case the queue is returned to ieee80211_return_txq
when iterating over the scheduled TXQs in mt76_txq_schedule_list.

This has the effect of inserting said TXQ at the head of the list. In
return this means the same TXQ as there are still frames pending which
AQL still considers transmittable. The TXQs following thus never get
scheduled in the iteration.

This can manifest in high latency, and low throughput for other clients.
Check if the non-AQL packet count exceeds the limit and not return the
TXQ. Schedule all TXQs in case the non-AQL limit can be satisfied again.

Signed-off-by: David Bauer <mail@david-bauer.net>
---
 tx.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

--- a/tx.c
+++ b/tx.c
@@ -216,12 +216,25 @@ mt76_tx_status_check(struct mt76_dev *dev, struct mt76_wcid *wcid, bool flush)
 EXPORT_SYMBOL_GPL(mt76_tx_status_check);
 
 static void
 mt76_tx_check_non_aql(struct mt76_dev *dev, struct mt76_wcid *wcid,
 		      struct sk_buff *skb)
 {
 	struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
+	struct ieee80211_sta *sta;
 	int pending;
+	int i;
 
 	if (!wcid || info->tx_time_est)
 		return;
 
 	pending = atomic_dec_return(&wcid->non_aql_packets);
 	if (pending < 0)
 		atomic_cmpxchg(&wcid->non_aql_packets, pending, 0);
+
+	sta = wcid_to_sta(wcid);
+	if (!sta || pending != MT_MAX_NON_AQL_PKT - 1)
+		return;
+
+	for (i = 0; i < ARRAY_SIZE(sta->txq); i++) {
+		if (!sta->txq[i])
+			continue;
+
+		ieee80211_schedule_txq(dev->hw, sta->txq[i]);
+	}
 }
